[FX] Add call_module to delegate

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44757 [FX][WIP] Add call_module to delegate**

Differential Revision: [D23719781](https://our.internmc.facebook.com/intern/diff/D23719781)

Commit:
[FX][WIP] Add call_module to delegate

[ghstack-poisoned]

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -209,11 +227,8 @@ def symbolic_trace(root : torch.nn.Module, delegate_class=DefaultDelegate) -> Gr
-        else:
-            target = _find_module(root, mod)
-            return _create_proxy(delegate, 'call_module', target, args, kwargs)
+        target = _find_module(root, mod)
```
I guess we can void passing orig_call by unpatching `Module.__call__` for the duration of `call_module`.  Then you can must `m(*args, **kwargs)`

jamesr66a replied:
Hm, I tried that but I think we actually need to make this more reentrant. When `call_module` calls `m`, inside that call the `module_call_wrapper` needs to be reinstated. Otherwise nothing beyond the first level of calls will go through `module_call_wrapper`

Commit:
Update on "[FX][WIP] Add call_module to delegate"


Differential Revision: [D23719781](https://our.internmc.facebook.com/intern/diff/D23719781)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Add call_module to delegate"


Differential Revision: [D23719781](https://our.internmc.facebook.com/intern/diff/D23719781)

[ghstack-poisoned]

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -135,11 +153,25 @@ def trace(self, root: torch.nn.Module) -> GraphModule:
+                            frame.frame.f_locals.get('disable_intercept', False):
+                        return mod._call_impl(*args, **kwargs)
+                    else:
+                        break
```
@zdevito  i got creative here. Let me know if there's a more sane way

dzhulgakov replied:
another creative idea: set `__call__` on **Module instance** for the duration of `model_call_wrapper` and then undo it back. That way you have two-level override :)

jamesr66a replied:
That doesn't work, it was the first thing I tried

jamesr66a replied:
Let me see if I can explain the different alternatives:
1) Patch something up in Tracer.call_module. This isn't ideal because any alternative implementations of `call_module` will need to know about this, and if they miss it they'll get weird infinite recursion errors
2) Double override like you're describing. This is fine except for when `mod` is recursive. In that case, we will lose all interception of calls except for the outermost recursive call. This can be fixed by patching something up in `Tracer.call_module`, but then we run into the same issues as (1)
3) To alleviate the common theme in (1) and (2), we could wrap `mod` in some object that will do this patching/unpatching business automatically. However, then `mod` doesn't work in `isinstance` checks
4) Directly implement this dispatch mechanism (my approach here). Basically I'm thinking about this as multi-dispatch, with an outer level for tracing, exactly like the PT dispatcher. This has been proven to work for all sorts of nested and recursive callsites, starting a new dispatch chain for each invocation. I've basically hacked this together with the `inspect` stuff, stripping away functionality if we've already done it (i.e 1 level dispatch key)

zdevito replied:
Looking at frame information here is too complicated for what it accomplishes. Seems like it would be possible to use a 'state' inside the module_call_wrapper to track where it is being invoked from. When 'tracing' it would do its tracing behavior and invoke `call_module`, but when it invokes `call_module` it would switch the state to 'passthrough', however, when invoked in the 'passthrough' state, it would switch back to 'tracing' during the actual module call. 

I think, though, that this is getting more complicated than it deserves. We don't have a use case yet where `call_module` needs to be completely custom. Our uses cases so far such as structure-preserving tracing just need to do some special things about `is_module_leaf`. So maybe we should just not implement this yet.

jamesr66a replied:
@zdevito I have heard some demand for having the `target` available in `is_leaf_module`, though, which this solves indirectly. I can put out a patch that just adds that directly

Commit:
Update on "[FX][WIP] Add call_module to delegate"


Differential Revision: [D23719781](https://our.internmc.facebook.com/intern/diff/D23719781)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -47,6 +47,24 @@ class Tracer(TracerBase):
+        preserve the original module hierarchy.
+        """
+        if not self.is_leaf_module(m):
+            return m(*args, **kwargs)
```
can't we just unpatch the `Module.__call__` before calling `Delegate.call_module` and here add something that patches it back:
```
with EmbedInline():
  m(*args, **kwargs)
```

or

```
embed_inline(m)(*args, **kwargs)
```

jamesr66a replied:
This method is overridable by the caller and we want to avoid them having to know the implementation details of how our weird module call patching thing works if they want to create their own implementation

jamesr66a replied:
In particular, it would be surprising if the user has a `Module` `Foo`, they see that `isinstance(m, Foo)`, so it would make sense that an instance of their own module would behave like they expect it to (i.e. `m(*args, **kwargs)` calls the module, not something else)
